### PR TITLE
feat(platform): add opt-in in-cluster east-west encryption toggle

### LINKED ISSUE #2977: Opt-in CNI transparent encryption for in-cluster (east-west) traffic
## Context

Cozystack ships several networking datapaths as `cozystack.networking` variants, and the default `isp-full` (Talos) bundle deploys **KubeOVN + Cilium** (`kubeovn-cilium`), not plain Cilium (`packages/core/platform/templates/bundles/system.yaml:94`). Each datapath has its own in-cluster encryption mechanism, and all are off today:

- **`kubeovn-cilium` (default)** — KubeOVN can encrypt the Geneve overlay with IPsec, but `ENABLE_OVN_IPSEC` defaults to `false` (`packages/system/kubeovn/charts/kube-ovn/values.yaml:78`).
- **`cilium` / `cilium-generic`** — Cilium 1.19.3 supports transparent WireGuard/IPsec encryption, but it is unset in `packages/system/cilium/values.yaml`.
- **`cilium-kilo`** — Kilo establishes a WireGuard mesh, but at `meshGranularity: location` it encrypts only cross-location traffic; pod-to-pod within a location stays in cleartext by design (`packages/system/kilo/values.yaml:7`).

In every shipped configuration, intra-location / intra-cluster pod-to-pod traffic is unencrypted by default, and there is no single, consistent operator-facing way to turn in-cluster encryption on.

The per-app TLS series (#2814) and Gateway TLS-passthrough (#2815) secure specific, explicitly-exposed endpoints; they do nothing for the bulk of east-west traffic between workloads that do not implement TLS. #2816 mentions CNI mesh encryption only as an optional complement; this issue carves it into an explicit, self-contained opt-in feature.

## Goal

A platform-level opt-in that enables in-cluster transport encryption for the active datapath, off by default. When enabled, pod-to-pod traffic across nodes is encrypted at the network layer regardless of whether the workload implements TLS — defense-in-depth for the whole mesh without per-application changes.

## Design questions

- Per-datapath knob vs a single platform abstraction. The mechanism differs by variant: KubeOVN `ENABLE_OVN_IPSEC`, Cilium native WireGuard/IPsec, Kilo WireGuard. A unified `networking.encryption.enabled` toggle would have to map onto whichever datapath is active.
- Default datapath first. Since `kubeovn-cilium` is the default, the KubeOVN IPsec path is the highest-value target; the Cilium-native path matters for the `cilium` / `cilium-generic` variants.
- Relationship to Kilo. Kilo already covers cross-location; this feature targets the intra-location gap. Decide whether they compose or are mutually exclusive per variant.
- WireGuard vs IPsec trade-offs where both are available; MTU / throughput overhead; interaction with the already-enabled host firewall and kube-proxy replacement.

## Non-goals

- Replacing application-level TLS or the per-app certificate series (#2814). In-cluster encryption does not reach an external client and does not give a tenant a verifiable endpoint certificate; it complements end-to-end TLS, it does not substitute for it.
- A fully CNI-agnostic abstraction that hides the datapath — the mechanism is necessarily datapath-specific, and surfacing that is acceptable.

## Acceptance

- A platform value enables in-cluster encryption for at least the default `kubeovn-cilium` datapath, defaulting to off.
- Documentation states which datapath each mechanism covers (incl. Kilo's cross-location-only scope), the measured overhead, and explicitly that this complements rather than replaces application-level / end-to-end TLS.
- The feature is verified working on the `isp-full` (Talos) variant on a live Talos cluster, not by static checks alone: the kube-ovn-cni pod starts cleanly (no CreateContainerError from a read-only-rootfs hostPath), IPsec ESP SAs establish between nodes, and encrypted pod-to-pod traffic flows across nodes. #2984 is not considered done until this live-on-Talos verification passes.

Part of #2811.

### PR DESCRIPTION
**Pull request by lexfrei**
## What this PR does

Adds `networking.encryption.enabled` to the platform chart — a single operator-facing opt-in, off by default, that turns on in-cluster (east-west) transport encryption for the active datapath. When enabled, pod-to-pod traffic across nodes is encrypted at the network layer regardless of whether the workload implements TLS: defense-in-depth for the whole mesh without per-application changes.

The default `isp-full` bundle selects the `kubeovn-cilium` datapath (and `isp-full-generic` selects `kubeovn-cilium-generic`), where KubeOVN owns the datapath and the overlay is a Geneve tunnel. The toggle therefore maps to KubeOVN's overlay IPsec (`kube-ovn` `ENABLE_OVN_IPSEC`). KubeOVN's own controller provisions the IPsec CA and per-node certificates through the Kubernetes CSR API, so no cert-manager wiring is needed.

This complements, and does not replace, application-level / end-to-end TLS: in-cluster encryption never reaches an external client and gives a tenant no verifiable endpoint certificate. The values comments document the per-datapath scope — including that Kilo's WireGuard mesh (`meshGranularity: location`) covers only cross-location traffic and leaves the intra-location gap this toggle targets, and that the Cilium-native WireGuard/IPsec path on the `cilium` / `cilium-generic` variants is not reached because the system bundle does not select those variants — plus the MTU/throughput overhead.

Scope: only the default KubeOVN datapath is wired through the platform toggle, which is what the system bundle actually selects. Wiring the Cilium-native path would require the bundle to expose cilium-only variant selection; that is out of scope here and called out in the values comments.

Validation: `helm unittest` on the platform chart passes (66 tests, 7 new) covering on / off / absent / null and verifying the IPsec flag deep-merges without dropping the existing `kube-ovn` overrides, for both `isp-full` and `isp-full-generic`.

Closes #2977

Part of #2811

### Screenshots

N/A — no UI changes.

### Release note

```release-note
feat(platform): add opt-in `networking.encryption.enabled` (off by default) that enables in-cluster east-west transport encryption for the default KubeOVN datapath via overlay IPsec; complements, and does not replace, application-level TLS.
```

## Summary by CodeRabbit

* **New Features**
  * Introduced opt-in in-cluster network encryption using KubeOVN IPsec for east-west traffic protection.
  * Added `networking.encryption` configuration toggle (disabled by default) with full backward compatibility.

* **Tests**
  * Added comprehensive test coverage for network encryption configurations across system variants and edge cases.

### CODE REVIEW

File: packages/core/platform/values.yaml [hunk 1]
```diff
@@ -45,6 +45,44 @@ networking:
...
+  # CAVEATS: encrypting the overlay adds per-packet ESP overhead and lowers
+  # the effective MTU, so east-west throughput drops and latency rises; the
+  # cost is highest for high-bandwidth flows. IPsec here protects the
+  # inter-node overlay tunnels — same-node pod-to-pod traffic never leaves the
+  # host and is not tunneled or encrypted. Flipping this on an existing
+  # cluster triggers a rollout of the CNI components. On variants without a
+  # cozystack-managed encryptable datapath (e.g. isp-hosted / noop) the toggle
+  # has no effect.
+  encryption:
+    enabled: false
```
[the remark this answers is not in the document]
Comment (lexfrei) [reply packages/core/platform/values.yaml:85]: Thanks. I checked this against the package and there's nothing to regenerate here. `packages/core/platform` has no `values.schema.json` (no `core/*` package does), and its `Makefile` has no `generate` target, so `make generate -C packages/core/platform` errors rather than emitting a schema. The `values.schema.json` / cozyvalues-gen pipeline is for the user-facing catalog charts: the `run-make-generate` pre-commit hook iterates `./packages/apps/*/` and `./packages/extra/*/` only, never `core/`, and the repo-root `make generate` (codegen-drift) is scoped by its `files:` filter to the Go API paths. So no schema gate applies to this platform-chart change; the operator-facing docs for the new key live in the `values.yaml` comments. Leaving as-is.

**Review by myasnikovdaniil — APPROVED** [review]
LGTM — approving. Clean, purely-additive opt-in with the safety properties this kind of toggle needs.

Verified locally (render/static — did not enable IPsec on any live cluster):

- **Default-off is byte-identical to `main`.** Rendered `templates/bundles/system.yaml` with the toggle absent on this branch vs `main` → zero diff, for `isp-full`. Existing installs are untouched.
- **On-delta is exactly the IPsec flag, deep-merged.** Toggle on → the only change is `func.ENABLE_OVN_IPSEC: true` nested under `kube-ovn`, with the existing `ipv4` overrides preserved (not replaced). Confirmed for both `isp-full` (kubeovn-cilium) and `isp-full-generic` (kubeovn-cilium-generic); no-op on `isp-hosted`/noop as documented.
- **Wiring is complete, not a dead flag.** `func.ENABLE_OVN_IPSEC` is consumed by the vendored kube-ovn chart (`--enable-ovn-ipsec` on both controller and CNI daemonset, IPsec key-dir mount, `runAsUser: 0`), and the controller self-manages the IPsec CA + per-node certs via the K8s CSR API (RBAC for `certificatesigningrequests` approve/sign + `kubeovn.io/signer` present) — so the "no cert-manager wiring" claim holds.
- Variant scoping in the values comments is accurate: the system bundle only selects `kubeovn-cilium`/`kubeovn-cilium-generic`/`noop`, so the Cilium-native and Kilo notes are correct. Kilo's `meshGranularity: location` default checks out.
- `helm unittest` 66/66 pass (7 new), `helm lint` clean, `core/platform` is not cozyvalues-gen managed so no schema regen needed.

One **optional** hardening thought (non-blocking): enabling east-west IPsec on a running cluster is cluster-wide-disruptive (CNI rollout, ovs restart, effective-MTU drop) with no clean live disable. The values comment documents this well, so an operator flipping it via GitOps is informed. But this is exactly the class of change the `proxyProtocol` hazard guard a few lines down in the same file was built to gate. Worth considering whether a similar render-time acknowledgement (e.g. a `networking.encryptionAcknowledgeDisruption: true` confirm) is warranted, or whether the doc comment is sufficient given there's no hidden state asymmetry here (unlike proxyProtocol). Happy to merge as-is either way.

**Comment by lexfrei** [comment]
Heads-up from testing this on a live Talos cluster: with `ENABLE_OVN_IPSEC=true` (what the `networking.encryption.enabled` toggle drives on the kubeovn-cilium datapath), the `kube-ovn-cni` pod fails to start on Talos nodes.

The IPsec path mounts a hostPath for the OVS IPsec keys at `/etc/origin/ovs_ipsec_keys`. On Talos `/etc/origin` is a read-only filesystem, so the kubelet can't create the directory:

```
failed to generate container spec: failed to apply OCI options: failed to mkdir "/etc/origin/ovs_ipsec_keys": mkdir /etc/origin: read-only file system
```

The container never starts (`CreateContainerError`), so no IPsec SAs establish (0 ESP SAs) and the overlay datapath is degraded on every Talos node. This affects the `isp-full` variant, which is the default/primary platform.

Static checks don't surface this — helm-unittest and chart review both pass, because the template renders correctly; the failure is a runtime hostPath-vs-read-only-rootfs issue specific to Talos that only appears on a live Talos node.

Fix direction: the IPsec keys path needs to be Talos-writable — a hostPath under `/var` (writable on Talos) or an emptyDir — rather than `/etc/origin`. Holding this in draft until that's addressed.

**Comment by lexfrei** [comment]
Follow-up from the same Talos cluster, with the keys path fixed: the direction above was right and the knob exists. `OVN_IPSEC_KEY_DIR` is a kube-ovn value, and we already redirect its siblings (`OPENVSWITCH_DIR`, `OVN_DIR`) to `/var/lib` but never set the IPsec one, so it fell back to the read-only `/etc/origin` default. Setting `OVN_IPSEC_KEY_DIR: /var/lib/ovs_ipsec_keys` makes `kube-ovn-cni` start cleanly.

That half is really an upstream documentation gap, and I sent it there: the Talos sections of both kube-ovn charts list only two of the three host directories and leave the IPsec one at its `/etc/origin` default, so following the documented instructions and then enabling IPsec reproduces the error above. [kubeovn/kube-ovn#7041](https://github.com/kubeovn/kube-ovn/pull/7041) adds the missing directory to both sections, together with a render fix it turned up along the way — the `ovs-ipsec-keys` volume is indented one level too deep, which makes the v2 chart fail to render at all whenever `enableOvnIpsec` is set.

That unblocks container creation, but not the feature. With it applied IPsec does come up — 8 ESP SAs establish between node IPs in transport mode, and notably they establish with `cilium.hostFirewall.enabled=true`, so the host firewall does not block SA establishment. Two blockers remain, and both are why this stays in draft:

1. **Cross-node pod-to-pod traffic is 100% lost while IPsec is on.** The SAs are up, but the encrypted datapath carries nothing. I have not isolated it further; the candidates are ESP (IP proto 50) dropped on the underlay, MTU/fragmentation, or a kube-ovn IPsec datapath bug on this setup. This is distinct from SA establishment, which demonstrably works.
2. **There is no clean teardown.** Turning IPsec back off leaves stale xfrm SAs and policies behind — kube-ovn does not flush them on disable — so cross-node pod-to-pod stays broken after the rollback until `ip xfrm state flush` and `ip xfrm policy flush` are run by hand on every node. That makes the toggle a one-way door in practice.

So the blocker is neither the chart wiring nor the keys path — both are solved. It is that kube-ovn IPsec does not carry traffic on this datapath and cannot be turned off cleanly, which looks like an upstream problem rather than something this toggle can paper over. Unlike the keys path, these two are not reported upstream yet. I am keeping this in draft until the datapath and the teardown are fixed: shipping an opt-in that silently blackholes pod-to-pod traffic and cannot be reverted without manual surgery on every node is not worth the defense-in-depth.
